dds/DCPS/TypeAssignability.cpp

```diff
@@ -0,0 +1,2083 @@
+   *        The second type can be anything.
+   */
+  bool TypeAssignability::assignable_annotation(const MinimalTypeObject& ta,
+                                                const TypeIdentifier& tb) const
+  {
+    // No rule for annotation in the spec
+    return false;
+  }
+
+  /**
+   * @brief The first type must be TK_STRUCTURE.
+   *        The second type must not be TK_ALIAS.
+   */
+  bool TypeAssignability::assignable_struct(const MinimalTypeObject& ta,
+                                            const MinimalTypeObject& tb) const
+  {
+    if (TK_STRUCTURE != tb.kind) {
+      return false;
+    }
+
+    // Extensibility kind must match
+    TypeFlag extensibility_mask = IS_FINAL | IS_APPENDABLE | IS_MUTABLE;
+    if ((ta.struct_type.struct_flags & extensibility_mask) !=
+        (tb.struct_type.struct_flags & extensibility_mask)) {
+      return false;
+    }
+
+    // Any members in T1 and T2 that have the same name also have
+    // the same ID, and vice versa
+    MatchedSet matched_members;
+    for (size_t i = 0; i < ta.struct_type.member_seq.members.size(); ++i) {
+      MemberId id_a = ta.struct_type.member_seq.members[i].common.member_id;
+      const NameHash& h_a = ta.struct_type.member_seq.members[i].detail.name_hash;
+      ACE_CDR::ULong
+        name_a = (h_a[0] << 24) | (h_a[1] << 16) | (h_a[2] << 8) | (h_a[3]);
+      for (size_t j = 0; j < tb.struct_type.member_seq.members.size(); ++j) {
+        MemberId id_b = tb.struct_type.member_seq.members[j].common.member_id;
+        const NameHash& h_b = tb.struct_type.member_seq.members[j].detail.name_hash;
+        ACE_CDR::ULong
+          name_b = (h_b[0] << 24) | (h_b[1] << 16) | (h_b[2] << 8) | (h_b[3]);
+        if ((name_a == name_b && id_a != id_b) ||
+            (id_a == id_b && name_a != name_b)) {
+          return false;
+        } else if (name_a == name_b && id_a == id_b) {
+          matched_members.push_back(std::pair<const MinimalStructMember*,
+                                    const MinimalStructMember*>
+                                    (&ta.struct_type.member_seq.members[i],
+                                     &tb.struct_type.member_seq.members[j]));
+          break;
+        }
+      }
+    }
+
+    // There is at least one member m1 of T1 and one corresponding member
+    // m2 of T2 such that m1.id == m2.id
+    if (matched_members.size() == 0) {
+      return false;
+    }
+
+    // For any member m2 of T2, if there is a member m1 of T1 with the same
+    // ID, then the type KeyErased(m1.type) is-assignable-from the type
+    // KeyErased(m2.type)
+    for (size_t i = 0; i < matched_members.size(); ++i) {
+      const CommonStructMember& member = matched_members[i].second->common;
+      const MinimalTypeObject* toa = 0;
+      const MinimalTypeObject* tob = 0;
+      bool type_matched = false;
+      // KeyErased(T) is defined only for struct and union
+      if (get_struct_member(&tob, member)) {
```

Comment:
Does the 1st parameter type really need to be a `**` or could it be `*&`?

Reply:
Yes, `*&` works too. I've updated to use `*&`.